### ISSUE TITLE
An unexpected error occurred trying to execute the MRefBuilder MSBuild task: System.InvalidOperationException

### ISSUE THREAD

**Issue by wollmich** [EWSoftware/SHFB#1006]
I'm using:

- Windows 10 Enterprise 22H2 (64 bit)
- Visual Studio 2019 Professional
- .NET Framework V4.6.2
- Sandcastle Help File Builder and Tools Version 2023.3.4.0

The following minimal example:

```
namespace SandCastleBug
{
    /// <summary>
    /// Interface Number
    /// </summary>
    /// <typeparam name="T">Type</typeparam>
    public interface INumber<T>
        where T : INumber<T>
    {
    }

    /// <summary>
    /// Interface Real Number
    /// </summary>
    /// <typeparam name="T">Real Type</typeparam>
    public interface IRealNumber<T> : INumber<T>
        where T : IRealNumber<T>
    { 
    }

    /// <summary>
    /// Interface Complex Number
    /// </summary>
    /// <typeparam name="T1">Complex Type</typeparam>
    /// <typeparam name="T2">Real Type</typeparam>
    public interface IComplexNumber<T1, T2> : INumber<T1>
        where T1 : IComplexNumber<T1, T2>
        where T2 : IRealNumber<T2>
    {
    }

    /// <summary>
    /// Interface Array
    /// </summary>
    /// <typeparam name="T">Array Type</typeparam>
    /// <typeparam name="D">Element Type</typeparam>
    public interface INArray<T, D>
        where T : INArray<T, D>
        where D : INumber<D>
    {
        /// <summary>
        /// Initializes a 3D-Array
        /// </summary>
        /// <param name="data">Data</param>
        void Init(D[, ,] data);
        /// <summary>
        /// Initializes a 3D-Array
        /// </summary>
        /// <param name="data">Data</param>
        void Init(D[][][] data);
    }
    
    /// <summary>
    /// Interface Real Array
    /// </summary>
    /// <typeparam name="T">Real Array Type</typeparam>
    /// <typeparam name="D">Real Element Type</typeparam>
    public interface IRealNArray<T, D> : INArray<T, D>
        where T : IRealNArray<T, D>
        where D : IRealNumber<D>
    {
    }
    
    /// <summary>
    /// Interface Complex Array
    /// </summary>
    /// <typeparam name="T1">Complex Array Type</typeparam>
    /// <typeparam name="T2">Real Array Type</typeparam>
    /// <typeparam name="D1">Complex Element Type</typeparam>
    /// <typeparam name="D2">Real Element Type</typeparam>
    public interface IComplexNArray<T1, T2, D1, D2> : INArray<T1, D1>
        where T1 : IComplexNArray<T1, T2, D1, D2>
        where T2 : IRealNArray<T2, D2>
        where D1 : IComplexNumber<D1, D2>
        where D2 : IRealNumber<D2>
    {
    }
}
```

causes the following error:

```
-------------------------------
[Sandcastle Help File Builder and Tools, version 2023.3.4.0]
Creating output and working folders...
The Sandcastle tools are located in 'C:\Program Files (x86)\EWSoftware\Sandcastle Help File Builder'
Locating components in the following folder(s):
    H:\CSharp\SandCastleBug
    C:\ProgramData\EWSoftware\Sandcastle Help File Builder\Components and Plug-Ins
    C:\Program Files (x86)\EWSoftware\Sandcastle Help File Builder\Components
Using framework reflection data for '.NET Framework 4.6.2' located in 'C:\Program Files (x86)\EWSoftware\Sandcastle Help File Builder\Data\.NETFramework'
Using presentation style 'Default2022' located in 'C:\Program Files (x86)\EWSoftware\Sandcastle Help File Builder\Components\Default2022'
Loading and initializing plug-ins...
Website Table of Contents Generator Version 2023.3.4.0
Copyright © 2006-2023, Eric Woodruff, All Rights Reserved
-------------------------------
Clearing working folder...
-------------------------------
Clearing any prior web/markdown output...
-------------------------------
Validating and copying documentation source information
Source: H:\CSharp\SandCastleBug\SandCastleBug.csproj
    Found project 'H:\CSharp\SandCastleBug\SandCastleBug.csproj'

Parsing project files
    Found assembly 'H:\CSharp\SandCastleBug\bin\Release\SandCastleBug.dll' (.NETFramework 4.6.2)

References to include:
    None

Copying XML comments files
    H:\CSharp\SandCastleBug\bin\Release\SandCastleBug.xml -> H:\CSharp\SandCastleBug\Help\Working\SandCastleBug.xml
-------------------------------
Generating shared content files (en-US, English (United States))...
    Last step completed in 00:00:00.0220
-------------------------------
Generating API filter for MRefBuilder...
    Last step completed in 00:00:00.0060
-------------------------------
Generating reflection information...
[C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Current\Bin\MSBuild.exe - GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
Build started 26.06.2023 16:17:20.
Project "H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj" on node 1 (default targets).
PrepareForBuild:
  Creating directory "obj\Debug\".
GenerateRefInfo:
  Microsoft.Build (v16.11.2+f3225964253d565862806752e54a7608191a9035)
  ¸ Microsoft Corporation. All rights reserved.
  No source code context base path specified.  Source context information is unavailable.
  Loaded 1 assemblies for reflection and 1 dependency assemblies.
MRefBuilder : error : An unexpected error occurred trying to execute the MRefBuilder MSBuild task: System.InvalidOperationException: SandCastleBug.INArray`2<type parameter.T1,type parameter.D1> [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error : SandCastleBug.INArray`2<type parameter.T1,type parameter.D1>.Init(type parameter.D1[][][]) [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error : No members in the template matched the parameters of the specialization.  This is not possible but apparently it happened. [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.Reflection.ReflectionUtilities.GetTemplateMember(Member member) [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.ManagedReflectionWriter.WriteTypeElements(TypeNode type) [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.ManagedReflectionWriter.WriteType(TypeNode type) [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.ManagedReflectionWriter.VisitType(TypeNode type) [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.Reflection.ApiVisitor.VisitTypes(TypeNodeList types) [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.Reflection.ApiVisitor.VisitNamespace(Namespace space) [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.ManagedReflectionWriter.VisitNamespace(Namespace space) [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.Reflection.ApiVisitor.VisitNamespaces(NamespaceList spaces) [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.ManagedReflectionWriter.VisitNamespaces(NamespaceList spaces) [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.Reflection.ApiVisitor.VisitApis() [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.MSBuild.MRefBuilder.GenerateReflectionInformation() [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
MRefBuilder : error :    at Sandcastle.Tools.MSBuild.MRefBuilder.Execute() [H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj]
Done Building Project "H:\CSharp\SandCastleBug\Help\Working\GenerateRefInfo-4bfcf7ae-1acb-4157-9435-6adbed7ecc4c.proj" (default targets) -- FAILED.
    Last step completed in 00:00:01.4831
-------------------------------

SHFB: Error BE0043: Unexpected error detected in last build step.  See build log for details.
   at SandcastleBuilder.Utils.BuildEngine.TaskRunner.Run(String processFilename, String targetFile, String arguments)
   at SandcastleBuilder.Utils.BuildEngine.BuildProcess.Build()
```

The above minimal example worked (.NET target framework version was 4.5) with Sandcastle Help File Builder and Tools Version 2015.1.12.0.
I've tested as well Version 2016.9.17.0 and there it fails.

I've attached the source and project files.
[SandCastleBug.zip](https://github.com/EWSoftware/SHFB/files/11870478/SandCastleBug.zip)

**Comment by EWSoftware** [comment]
Fixed along with #994

**Comment by wollmich** [comment]
I've tested it today with the new release 2023.7.8.0 and it works now. Thanks for the fast bug fixed and the new release.
